### PR TITLE
Typed Events for the Indexer Event Handler

### LINKED ISSUE #113: Typesafe Event Handler

### PR DESCRIPTION
**Pull request by boudra**
closes #113

The idea of this PR is to reduce typos and mistakes when accessing fields in the events or fetching fields from the contracts. It implements the following:

- Removes all eslint disable comments and implicit / explicit any
- Full types for the indexer entities (rounds, applications, votes etc.)
- Full types for the events that are handled, to know what arguments are available for each event.
- A typed round contract

I opted to do the types with plain Typescript instead of leveraging Viem, ABIType or Typechain because it was easier to implement and to be honest I don't think we have that many events to worry about automating this stuff. But we can think of doing it in the future, especially when we migrate to Allo v2.

### CODE REVIEW

**Review by bhargavaparoksham — APPROVED** [review]
lgtm.

**Comment by bard** [comment]
It might be worth considering a version that, although less sophisticated, is a tad more common and spares brain cycles. Something like:

```typescript
interface NewProjectApplicationEvent extends ChainsauceEvent {
  name: "NewProjectApplication";
  projectID: ethers.BigNumber;
  project: string;
  applicationIndex: ethers.BigNumber;
  applicationMetaPtr: MetaPtr;
}

interface ApplicationStatusesUpdatedEvent extends ChainsauceEvent {
  name: "ApplicationStatusesUpdated"
  index: ethers.BigNumber;
  status: ethers.BigNumber;
}

export type Event =
  | NewProjectApplicationEvent
  | ApplicationStatusesUpdatedEvent
```

There's some redundancy there (the `name` property repeats the interface name) and chance for mistake (one might add an interface and forget the union) but the former is cosmetic and the latter is apparent at compile time, so I'm usually happy to pay the price to help the reader.

That said, it's functionally equivalent (conditionals narrow the type), so the importance is relative.

The other thing I'd want to consider is, since we might switch to a type generator later, how similar is this to what the type generator would produce? If we handcraft something now that's close enough to that, the transition will be smoother.

**Comment by boudra** [comment]
@bard good point!! I agree that we can roll with just a discriminated union, I'll make the updates 👍

**Review by vacekj — APPROVED** [review]

File: src/indexer/handlers/applicationMetaPtrUpdated.ts [hunk 1]
```diff
@@ -1,29 +1,36 @@
...
-    const metadata = await ipfs(metaPtr, cache);
+  const metaPtr = args.newMetaPtr.pointer;
+  const metadata = await ipfs<Round["applicationMetadata"]>(metaPtr, cache);
 
-    // eslint-disable-next-line @typescript-eslint/no-explicit-any
-    await db.collection("rounds").updateById(id, (round: any) => {
-      if (round.applicationMetaPtr === event.args.newMetaPtr.pointer) {
-        return { ...round, applicationMetadata: metadata };
-      }
+  if (!metadata) {
```
Comment (bard) [inline src/indexer/handlers/applicationMetaPtrUpdated.ts:25]: Is this kind of conditions expected or does it suggest something going wrong? If the latter, might warrant logging a warn-level message to stdout.

File: src/indexer/handlers/matchAmountUpdated.ts [hunk 2]
```diff
@@ -1,24 +1,29 @@
...
-  event: Event
+  event: MatchAmountUpdatedEvent
 ) {
   const id = event.address;
   const matchAmount = event.args.newAmount.toString();
 
-  const round = await db.collection("rounds").findById(id);
+  const round = await db.collection<Round>("rounds").findById(id);
+
+  if (!round) {
```
Comment (bard) [inline src/indexer/handlers/matchAmountUpdated.ts:15]: Same as above.

**Review by bard — DISMISSED** [review]

File: src/indexer/handlers/matchAmountUpdated.ts (hunk 2 above)
Comment (boudra) [reply src/indexer/handlers/matchAmountUpdated.ts:15]: This is an unexpected error, match amounts should always happen after a round is created. I have updated to throw an error.

File: src/indexer/handlers/applicationMetaPtrUpdated.ts (hunk 1 above)
Comment (boudra) [reply src/indexer/handlers/applicationMetaPtrUpdated.ts:25]: This one is expected, sometimes we get events with an invalid IPFS hash, the `ipfs` call returns undefined in those cases and we shouldn't error, there's a log inside that function.

File: src/indexer/handleEvent.ts [hunk 3]
```diff
@@ -31,22 +41,33 @@ function fullProjectId(
...
-    eventRenames[indexer.chainId]?.[event.address]?.[event.name] ?? event.name;
+    eventRenames[indexer.chainId]?.[originalEvent.address]?.[
+      originalEvent.name
+    ] ?? originalEvent.name;
 
-  switch (eventName) {
+  const event = {
+    ...originalEvent,
+    name: eventName,
+  } as Event;
```
[the remark this answers is not in the document]
Comment (boudra) [reply src/indexer/handleEvent.ts:57]: Do you have any idea on how to do this without a cast? or a safer alternative?

I'm essentially casting a type of:

```
interface UntypedEvent {
    name: string;
}
```

to a :

```
interface TypedEventA {
  name: "a";
}

interface TypedEventB {
  name: "b";
}

type TypedEvent = TypedEventA | TypedEventB;
```

[the remark this answers is not in the document]
Comment (boudra) [reply src/indexer/handleEvent.ts:57]: I guess checking the name at runtime is the safest way 😂

[the remark this answers is not in the document]
Comment (bard) [reply src/indexer/handleEvent.ts:57]: I had a quick try before commenting but couldn't sort it out. I don't think it's important right now, just a minor corner to smooth out at some point down the road.

[the remark this answers is not in the document]
Comment (boudra) [reply src/indexer/handleEvent.ts:57]: Thank you, ideally the event would come with a concrete type from Chainsauce so we wouldn't need to cast it here.

**Review by vacekj — APPROVED** [review]

**Review by bard — APPROVED** [review]

**Pull request merged**